### PR TITLE
Add handler for `SemanticsAction.scrollToOffset`

### LINKED ISSUE #159515: Clean up `SemanticsAction.scrollToOffset` TODOs
A new `SemanticsAction``SemanticsAction.scrollToOffset` is added to dart:ui. Some framework tests / lints that are making sure all semantics actions are handled and they had to be changed before the PR that introduces the new semantics action can be merged. Those PRs need to be updated / reverted after the engine roll.

https://github.com/flutter/flutter/pull/159108
https://github.com/flutter/flutter/pull/159407

(or search for TODOs with 159515 in its description)

### PR DESCRIPTION
**Pull request by LongCatIsLooong**
Also closes https://github.com/flutter/flutter/issues/159515

~Looking at the google testing failures~ cl/703169955

## Pre-launch Checklist

- [ ] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [ ] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [ ] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [ ] I signed the [CLA].
- [ ] I listed at least one issue that this PR fixes in the description above.
- [ ] I updated/added relevant documentation (doc comments with `///`).
- [ ] I added new tests to check the change I am making, or this PR is [test-exempt].
- [ ] I followed the [breaking change policy] and added [Data Driven Fixes] where supported.
- [ ] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#overview
[Tree Hygiene]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md
[test-exempt]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/blob/main/docs/contributing/Style-guide-for-Flutter-repo.md
[Features we expect every widget to implement]: https://github.com/flutter/flutter/blob/main/docs/contributing/Style-guide-for-Flutter-repo.md#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/blob/main/docs/contributing/Chat.md
[Data Driven Fixes]: https://github.com/flutter/flutter/blob/main/docs/contributing/Data-driven-Fixes.md

### CODE REVIEW

File: packages/flutter/lib/src/widgets/scrollable.dart [hunk 1]
```diff
@@ -611,12 +611,10 @@ class ScrollableState extends State<Scrollable> with TickerProviderStateMixin, R
...
     _configuration = widget.scrollBehavior ?? ScrollConfiguration.of(context);
-    _physics = _configuration.getScrollPhysics(context);
-    if (widget.physics != null) {
-      _physics = widget.physics!.applyTo(_physics);
-    } else if (widget.scrollBehavior != null) {
-      _physics = widget.scrollBehavior!.getScrollPhysics(context).applyTo(_physics);
-    }
+    final ScrollPhysics? physicsFromWidget = widget.physics ?? widget.scrollBehavior?.getScrollPhysics(context);
+    _physics =  _configuration.getScrollPhysics(context);
+    _physics = physicsFromWidget?.applyTo(_physics) ?? _physics;
```
Comment (chunhtai) [inline packages/flutter/lib/src/widgets/scrollable.dart:616]:
```suggestion
    _physics = physicsFromWidget?.applyTo(_physics) ?? _configuration.getScrollPhysics(context);
```

File: packages/flutter/lib/src/widgets/scrollable.dart [hunk 2]
```diff
@@ -1629,6 +1635,14 @@ class _RenderScrollSemantics extends RenderProxyBox {
     markNeedsSemanticsUpdate();
   }
 
+  void _onScrollToOffset(Offset targetOffset) {
+    final double offset = switch (axis) {
+      Axis.horizontal => targetOffset.dx,
```
Comment (chunhtai) [inline packages/flutter/lib/src/widgets/scrollable.dart:1640]: will there be cases where dx and dy both none zero?

**Review by chunhtai — APPROVED** [review]
LGTM

File: packages/flutter/lib/src/semantics/semantics.dart [hunk 3]
```diff
@@ -3920,6 +3924,28 @@ class SemanticsConfiguration {
...
+  /// The handler for [SemanticsAction.scrollToOffset].
+  ///
+  /// This handler is only called on iOS by UIKit, when the iOS focus engine
+  /// switches its focus to an item too close to a scrollable edge of a
+  /// scrollable container, to make sure the focused item is always fully
+  /// visible.
+  ///
+  /// The callback, if not `null`, should typically set the scroll offset of
+  /// the associated scrollable container to the given `targetOffset` without
+  /// animation as it is already animated by the caller: the iOS focus engine
```
[the remark this answers is not in the document]
Comment (LongCatIsLooong) [reply packages/flutter/lib/src/semantics/semantics.dart:3936]: UIKit would create a display link that calls `setContentOffset` every frame with different scroll offsets. So the animation is determined by UIKIt and the curve is determined by UIKit.

[the remark this answers is not in the document]
Comment (LongCatIsLooong) [reply packages/flutter/lib/src/semantics/semantics.dart:3936]: (should I rephrase somehow to make this clear?)

File: packages/flutter/lib/src/widgets/scrollable.dart (hunk 1 above)
Comment (LongCatIsLooong) [reply packages/flutter/lib/src/widgets/scrollable.dart:616]: hmm if I apply the suggested change, is `_physics` going to be up to date for the expression before `??`

File: packages/flutter/lib/src/semantics/semantics.dart (hunk 3 above)
[the remark this answers is not in the document]
Comment (chunhtai) [reply packages/flutter/lib/src/semantics/semantics.dart:3936]: yes, some rephrasing can help

Do we translate all the setContentOffset to scrollToOffset call?

File: packages/flutter/lib/src/widgets/scrollable.dart (hunk 1 above)
Comment (chunhtai) [reply packages/flutter/lib/src/widgets/scrollable.dart:616]: oh I didn't realize the applyTo uses _physics, in that case no, the suggestion is wrong, the original call is fine

File: packages/flutter/lib/src/widgets/scrollable.dart (hunk 2 above)
Comment (LongCatIsLooong) [reply packages/flutter/lib/src/widgets/scrollable.dart:1640]: Not currently. In the iOS implementation, every semantics node is only scrollable in one axis I think, so iOS won't issue commands that would scroll both axes. If it does then I think we can't do anything other than just ignoring the unscrollable axis?

Comment (chunhtai) [reply packages/flutter/lib/src/widgets/scrollable.dart:1640]: oh ok, just want to double check there isn't any surprise.

File: packages/flutter/lib/src/semantics/semantics.dart (hunk 3 above)
[the remark this answers is not in the document]
Comment (LongCatIsLooong) [reply packages/flutter/lib/src/semantics/semantics.dart:3936]: Yeah. Made some minor adjustments to the last sentence.

**Pull request merged**